ir/cid.go

```diff
@@ -2,16 +2,37 @@ package ir
 
 import (
 	"io"
+
+	"github.com/ipfs/go-cid"
 )
 
 type Cid struct {
```

Comment:
Type Cid is the "definition" of a cid smart tag.
You can generate a syntactic representation of the cid, using the method Dict() Dict.
Dicts are "syntactic representation".

You seem to have a misunderstanding. Your implementation of "Dict()" is trying to parse and validate a cid smart tag from cid.User. But its meant to work the other way around. cid.User is assumed to be correct. And Dict() simply adds the cid key to dict and tags it as "cid".

Your is valid, but belongs in a different place. You should revert the changes to type Cid, and just add your code
to a function called:

func ParseCidFromDict(d Dict) (Cid, error)

This will also resolve your issue with where to return the error.

Reply:
Oh I see! Now I get it! This is the syntactic representation after parsing. I completely misunderstood the code. It would be great if we can find some time next week for you to walk me through the code and how are you planning to structure it if you don't mind. Thanks!